### PR TITLE
implement bulk usermeta counterparts

### LINKED ISSUE #61: Add bulk usermeta endpoints
For now, the endpoint used to change usermeta only GET and SET a single usermeta.

https://github.com/j3k0/ganomede-users/#metadata

We will create bulk counterparts:

## /users/v1/auth/:token/multi/metadata/:keys [GET]

Users' protected custom data. `:keys` is a comma separated list of keys to retrieve.

### response [200] OK

    [{
        "key": "some-key",
        "value": "..."
    }, {
        "key": "some-other-key",
        "value": "..."
    }]

Uses this endpoint: https://github.com/j3k0/ganomede-usermeta/blob/master/README.md#get-authtokenkeys

## /users/v1/auth/:token/multi/metadata [POST]

Change multiple users' custom data.

Uses this endpoint: https://github.com/j3k0/ganomede-usermeta/blob/master/README.md#post-authtoken

### body (application/json)

    [{
        "key": "my-key",
        "value": "..."
    }, {
        "key": "my-other-key",
        "value": "..."
    }]

## /users/v1/multi/metadata/:userIds/:keys [GET]

Retrieve multiple users' custom data, using the users `id`.

Retrieve publicly available metadata. Both, `:userIds` and `:keys` are comma-separated list. Attach secret query string param to retrieve fields up to internal.

Missing or unknown keys, and those you are not allowed to read will be omitted (as opposed to being HTTP error).

Uses this endpoint: https://github.com/j3k0/ganomede-usermeta/blob/master/README.md#get-useridskeys

### response [200] OK

    [{
        "username": "alice",
        "key": "location",
        "value": "..."
    }, {
        "username": "bob",
        "key": "lastseen",
        "value": "..."
    }]

If the key doesn't exist or cannot be accessed, it will be omitted from the result.

---

See usermeta module.

### PR DESCRIPTION
**Pull request by HusseinTaha**
Closes #61

### CODE REVIEW

File: src/usermeta.ts [hunk 1]
```diff
@@ -31,20 +32,78 @@ export interface UsermetaClientOptions {
...
+        let oneLevelArray = (data as any[][])?.flat();
+        cb(null, oneLevelArray);
+      }
+    });
+  }
+  setBulk(pparams: UsermetaClientBulkOptions | string, keyValues: KeyValue[], cb: UsermetaClientBulkCallback) {
+    let tasks: any[] = [];
+    keyValues.forEach((kv) => {
+      tasks.push(cb2 => this['set'](pparams, kv.key, kv.value, cb2))
+    });
```
Comment (j3k0) [inline src/usermeta.ts:95]: For good style, don't write this:

```ts
let out = [];
in.forEach(x => out.push(...));
```

But write this:

```ts
const out = in.map(x => ...);
```

In the code above:

```ts
const tasks:Callback[] = keyValues.map((kv) => (cb2 => this['set'](pparams, kv.key, kv.value, cb2)));
```

File: src/usermeta.ts [hunk 2]
```diff
@@ -587,6 +690,112 @@ class UsermetaRouter {
...
+  getBulk(params: UsermetaClientBulkOptions | string, keys: string[], cb: UsermetaClientBulkCallback) {
+    //generate tasks for get.
+    if (keys.length === 0)
+      return cb(null, []);
+    const tasks: any[] = this.getTasksForBulk(params, keys, (_client, _keys) => {
+      let _tasks: any[] = [];
+      //if getBulk method exists on this client ..  then we call getbulk method => and passing the keys
+      //related.
+      if (_client.getBulk) {
+        _tasks.push(cb1 => _client.getBulk(params, _keys, cb1));
```
Comment (j3k0) [inline src/usermeta.ts:729]: return [cb1 => _client.getBulk(params, _keys, cb1)];

File: src/usermeta.ts [hunk 3]
```diff
@@ -587,6 +690,112 @@ class UsermetaRouter {
...
+      else {
+        //else we loop over the keys => and we will create tasks based for each key => client =>call for key.
+        _keys.forEach((key: string) => {
+          _tasks.push(cb1 => _client.get(params, key, (err: Error | null, reply?: string | null) => {
+            if (err) {
+              return cb1(err, reply);
+            }
+            cb1(err, { key, value: reply });
+          }));
+        });
```
Comment (j3k0) [inline src/usermeta.ts:740]:
```ts
          return _keys.map((key: string) => (
            cb1 => _client.get(params, key, (err: Error | null, reply?: string | null) => {
              if (err) {
                return cb1(err, reply);
              }
              cb1(err, { key, value: reply });
            }));
          ));
```

File: src/usermeta.ts [hunk 4]
```diff
@@ -587,6 +690,112 @@ class UsermetaRouter {
...
+      return _tasks;
+    });
+
+    async.parallel(tasks, (err, data) => {
+      if (err) {
+        cb(err, null);
+      }
+      else {
+        //make the 2 levels array to 1 level => [[], [], []] => []
+        let oneLevelArray = (data as any[][])?.flat();
```
Comment (j3k0) [inline src/usermeta.ts:751]: Here too, I hope we can get rid of this obscure `any[][]`. Especially as is goes into a variable (almost) called `array`, hard to understand what is in it...

File: src/usermeta.ts [hunk 5]
```diff
@@ -587,6 +690,112 @@ class UsermetaRouter {
...
+        _tasks.push(cb1 => _client.setBulk(pparams, values, cb1));
+      }
+      else {
+        //else we loop over the keys => and we will create tasks based for each key => client =>call for key/value.
+        _keys.forEach((key: string) => {
+          const val = keyValues.filter((x) => x.key == key)[0].value;
+          _tasks.push(cb1 => _client.set(pparams, key, val, cb1));
+        });
+      }
+      return _tasks;
```
Comment (j3k0) [inline src/usermeta.ts:777]: same remark, no need to create a variable called `_tasks` (see similar code above)

File: src/usermeta.ts [hunk 6]
```diff
@@ -31,20 +32,78 @@ export interface UsermetaClientOptions {
...
+        //make the 2 levels array to 1 level => [[], [], []] => []
+        let oneLevelArray = (data as any[][])?.flat();
+        cb(null, oneLevelArray);
+      }
+    });
+  }
+  setBulk(pparams: UsermetaClientBulkOptions | string, keyValues: KeyValue[], cb: UsermetaClientBulkCallback) {
+    let tasks: any[] = [];
+    keyValues.forEach((kv) => {
+      tasks.push(cb2 => this['set'](pparams, kv.key, kv.value, cb2))
```
Comment (j3k0) [inline src/usermeta.ts:94]: Why `this['set']` and not `this.set`? I think this is loosing the typing information...

Comment (HusseinTaha) [reply src/usermeta.ts:94]: in this section, its inside a class BulkedUsermetaClient, so this one doesn't implements or extends any other information. I needed some implementation that will be inherited into all other Clients types.

This class contains only the logic behind parallel and bulking

**Comment by j3k0** [comment]
A simple solution: pass a reference to a SimpleUsermetaClient  in
BulkedUsermetaClient  constructor. Store it as this.simpleClient  - then do
this.simpleClient.set(...) .

The cleaner (but more advanced) solution would be to make it an abstract
class. See
https://www.typescriptlang.org/docs/handbook/classes.html#abstract-classes -
add abstract set(...):void  to the declaration of BulkedUsermetaClient ).

**Comment by j3k0** [comment]
The non-deprecated link:
https://www.typescriptlang.org/docs/handbook/2/classes.html#abstract-classes-and-members

**Comment by HusseinTaha** [comment]
Please note that all comments are adjusted and pushed.

**Comment by j3k0** [comment]
Is it done? You can assign the PR to me when it's ready for review.

**Comment by HusseinTaha** [comment]
> Is it done? You can assign the PR to me when it's ready for review.

Yup it is, Sorry I forgot about the assign thing.

**Comment by j3k0** [comment]
@HusseinTaha I adjusted the test so they check for the correct values. Looks like there are some bugs:

- Local usermeta are not fetched (I think they are fetched from the Central usermeta repo, because both the local and central GanomedeUsermeta repo are the same "type".
- Directory usermeta are not returned with default values.

I also found something suspicious in `setBulk` (it's not using the `usernames` parameter), which means it doesn't need it maybe? Fixing this...

**Pull request merged**